packages/intrepid2/src/Discretization/Integration/Intrepid2_Cubature.hpp

```diff
@@ -152,7 +185,26 @@ namespace Intrepid2 {
                  weightViewType /* cubWeights */,
                  PointViewType  /* cellVertices */) const {
       INTREPID2_TEST_FOR_EXCEPTION( true, std::logic_error,
-                                    ">>> ERROR (Cubature::getCubature): this method should be over-riden by derived classes.");
+                                    ">>> ERROR (Cubature::getCubature): this method should be overridden by derived classes.");
+    }
+    
+    /** \brief Returns tensor cubature points and weights.  For non-tensor cubatures, the tensor structures are trivial, thin wrappers around the data returned by getCubature().  The provided containers should be pre-allocated through calls to allocateTensorCubaturePoints() and allocateTensorCubatureWeights().
+
+        \param cubPoints       [out]   - TensorPoints structure containing the cubature points.
+        \param cubWeights      [out]  - TensorData structure containing cubature weights.
+    */
+    virtual
+    void
+    getTensorCubature( const TensorPointDataType & tensorCubPoints,
```

Comment:
> Should we use getTensorCubature also for non-tensor elements?

Yes, I think so.  I think it makes sense to keep the existing `getCubature()` methods around for the sake of anyone who really wants to work directly with Kokkos containers rather than going through our TensorPoints container, but for most users, especially new users, I think it's better to use the new data structures, since they allow us to provide optimizations.

Do you think we should just call the new method `getCubature()` instead of `getTensorCubature()`?  It takes different types on input, so it would be a valid overload.  That might be clearer for users; I'm not sure.  Similarly for the `allocate()` methods.

Reply:
Good. Yes, I think it's better to overload `getCubature` and the allocate methods.

Reply:
This is now done.